### DISCUSSION TITLE
Implement per-channel fluid_synth_all_notes_off when seeking/stopping player

### DISCUSSION THREAD

**Discussion by albedozero** [FluidSynth/fluidsynth#981] [Ideas]
I'm working on an app that uses FluidSynth to play backing tracks from a MIDI file on some channels while the user can also trigger notes on other channels with an attached MIDI controller. The features added to fluid_player are great for this, particularly the ability to seek within a midi file, which makes it possible to set up loop points, jump to verse/chorus, etc. However, seeking [calls](https://github.com/FluidSynth/fluidsynth/blob/926581851ed1a095ef5b8659f77b38272d57e624/src/midi/fluid_midi.c#L2124) `fluid_synth_all_sounds_off(synth, -1)`, which creates an audible click as all playing sounds are cut off without being allowed to complete their release phase, and also stops any notes the user is playing on other MIDI channels.

I propose two things:
- Use `fluid_synth_all_notes_off` above instead, so that notes can release smoothly when seeking.
- Since `fluid_synth_all_notes_off` can already be called for specific channels, add a way of setting in the player which channels should receive an all_notes_off when seeking or stopping the player.

This involves adding an array of type `int` to `_fluid_player_t` and a function to the API for setting the array in a player object. I've created PR #980 that demonstrates/accomplishes this. The default behavior remains the same - if the user doesn't set the array, its initial value is `-1`, and all channels receive the all_notes_off call.

**Comment by derselbst** [comment]
Ok for me in general.

> Use fluid_synth_all_notes_off above instead, so that notes can release smoothly when seeking.

Pls. test how the behavior is when notes in the backing tracks are sustained: I.e. NoteOn, press sustain pedal, seek to a location where 
a) the sustain pedal is no longer depressed, and
b) the sustain pedal is still depressed.

There shouldn't be any hanging notes in a), but for b) I'm not entirely sure.

**Reply by mawe42** [reply]
As MIDI files contain only messages to alter synth state but not the state itself, I do wonder how this seeking is going to work reliably (even with the current `fluid_synth_all_notes_off()`. Do we go through all state changing messages (CCs, program changes, sysex etc) up to the point where we seek to and determine the right state before continuing with the file? If not, then I fear that concepts like 

> a) the sustain pedal is no longer depressed, and
> b) the sustain pedal is still depressed.

are not really meaningful.

**Reply by mawe42** [reply]
Ok, I could have read the source before commenting. It seems we do exactly that, go though the file from the beginning and play everything apart from note on/off until the point we seek to. Sorry for the noise.

**Reply by albedozero** [reply]
Can do @derselbst - I find there are hanging notes in b), also sometimes a), but also there are hanging notes if the player is _stopped_ when the sustain pedal is down, because it also calls `fluid_synth_all_notes_off` - this doesn't seem to be desired behavior. If there was a function that called `fluid_voice_release` for all the voices on the requested channel it seems like it would deal with the hanging notes, but since that's not the behavior of _All Notes Off_ or _All Sounds Off_ according to the soundfont spec (9.2), perhaps a function `fluid_synth_all_notes_release` could be added to _fluid_synth.c_?

Here's the rundown of what I found using this [test_mid.zip](https://github.com/FluidSynth/fluidsynth/files/7172313/test_mid.zip) file, included for reference (I could try to whip up some example C++ code too if that helps).

Seek from a point in the song where playing notes have been released (i.e. noteoff) but sustain pedal is depressed so that notes continue to sound, e.g. from tick 6200, to
a) point where sustain pedal has been released (tick 4200) - no hanging notes
b) point where sustain pedal is being held (tick 2300) - sustained notes continue to hang
c) point in song were sustain pedal is released but has never been pressed (CC64 has never changed - tick 0) - hanging notes
d) player is _stopped_ at any point when sustain pedal is down - hanging notes

@mawe42 your comment actually made me look at the code a bit closer to figure out what was going on in c), so noise appreciated :)

**Reply by albedozero** [reply]
With the latest commits to #980 adding `fluid_synth_all_notes_release`, the same tests as in my last reply give no hanging notes for a), b), or d). Seeking back to a point in the song before the sustain pedal has ever been pressed or released c) still produces hanging notes, because there is no CC64=0 message in the track up to that point for the player to use to set the sustain pedal state, so it remains at what it was. This seems like it would be rare/edge case however, and could be gotten around by programmatically sending the sustain pedal release message right after seeking, or putting a CC64=0 message in the MIDI file at the beginning of the song.

**Reply by mawe42** [reply]
The problem with c) could be handled by always issuing an "All Controllers Off" message before starting the seek. However... I think I remember us discussing if we should issue a controller reset between songs and decided against it. But I can't remember why, to be honest.

**Reply by derselbst** [reply]
I don't think that we need a `fluid_synth_all_notes_release`, because the client is free to "unpress" sustain or sostenuto pedal via `fluid_synth_cc`. Or ControllersOff as Marcus said. I'll try to find some time next weekend to have a closer listen to you provided example MIDI.

>  I think I remember us discussing if we should issue a controller reset between songs and decided against it. But I can't remember why, to be honest.

IIRC it would break seamless looping, or so. On the other hand, we could also simply risk it and see if somebody complains. But let's first investigate the "clicks". Then we can think about how the player should behave when stopping and when seeking.

**Comment by mawe42** [comment]
Just one more general thought about what you want to achieve. If you don't stop notes immediately before the seek but let them transition into their release phase and die out, you might get weird effects for sounds with a long-ish release phase, especially if they where affected by any controllers. Say a glockenspiel with a long release tail, which the song pitched down using a controller. You seek to a different position in the song, arrive at a different controller value and your (still audible) glockenspiel release phase suddenly jumps to a new pitch.

It seems to me that letting the sounds go into release instead of cutting them off opens up quite a few new edge cases.

**Reply by mawe42** [reply]
So for your two proposals, I would vote against the first one (letting notes release instead of cutting them off). In my opinion, it can't be one in a way that doesn't open up a lot of edge cases.

The second proposal I think is a really good idea. Only reset channels that the player has actually touched. For that you wouldn't even need a new API. We could let the player keep track of all channels that have received at least one MIDI event during playback. Then only reset those channels on seek / stop.

**Reply by mawe42** [reply]
As an alternative solution to the pops when stopping all notes: maybe we can implement a "softer" immediate stop? Bing the volume of all playing sounds down to 0 in a few milliseconds instead of simply halting their output. That would definitely be much more involved, but might satisfy your use-case without introducing unwanted edge cases.

**Reply by albedozero** [reply]
I see your point about the edge cases when using `fluid_voice_release` instead of `fluid_voice_off`. I think I can deal with the audio clicks for now by choosing loop points carefully, and perhaps in the future experiment with some kind of "quick release" version of `fluid_voice_off` as you say.

As for having the player keep track of the channels it has used and only resetting those on seek/stop - I agree that sounds a lot cleaner and I like that it wouldn't require adding to the API. One issue is that the player may output to a router which then re-routes messages to different channels (my app actually does this so the user doesn't have to reassign their controller's MIDI channel to avoid conflicting with the player) - in which case the player wouldn't know which channels to call `fluid_synth_all_sounds_off` for. However, it shouldn't have to - maybe what the player should be doing when stopping/seeking is generating an ALL_SOUND_OFF control change fluid_event on whatever channels it has touched, then any subsequent router(s) can send the event to the appropriate synth channel. I can change my PR to do that if it sounds like it would work?

**Reply by mawe42** [reply]
Sending ALL_SOUND_OFF and ALL_CTRL_OFF as MIDI messages instead of via the API should be possible, they are simply CC messages after all. The question is: do we want those important messages to go through the (possibly custom) MIDI callback and/or MIDI router? Personally I see no reason against it... it gives even more flexibility to the player. But it would be a change of behaviour and people might have implemented custom callbacks or routers that don't handle those messages properly.

Not sure... what do you think @derselbst?

**Reply by albedozero** [reply]
Maybe add a flag that would set the behavior of sending ALL_SOUND_OFF vs. calling fluid_synth_all_sounds_off?

**Reply by albedozero** [reply]
I updated the PR, adding a `player.global-mute` flag that is 1 by default and tells the player to call `fluid_synth_all_sounds_off` for all channels when seeking\stopping, and if 0 will send an ALL_SOUNDS_OFF event via the player's playback_callback only on the channels that have received notes. It works fine in my tests in either case, and doesn't require adding to the API (other than adding a fluidsetting). Any thoughts?

**Reply by derselbst** [reply]
> maybe we can implement a "softer" immediate stop? Bing the volume of all playing sounds down to 0 in a few milliseconds instead of simply halting their output

I would also like to investigate those "clicks" a little closer. This way, everyone who uses `fluid_synth_all_sounds_off` could benefit from a better implementation. Perhaps, Bill, you could provide an example audio file, so we can try to reproduce it on our side more easily. Ideally with MIDI and SF2 (in case it matters).

> Sending ALL_SOUND_OFF and ALL_CTRL_OFF as MIDI messages instead of via the API should be possible, they are simply CC messages after all. The question is: do we want those important messages to go through the (possibly custom) MIDI callback and/or MIDI router? Personally I see no reason against it... it gives even more flexibility to the player

Sending it as CC rather than directly via synth API actually seems more correct to me as well.  Yes, it could break some client - but I consider this pretty unlikely, so I would be fine with that. For the time we test it, you can keep the `player.global-mute` flag, Bill.

**Reply by albedozero** [reply]
Big oops on my part, Tom - the clicks were the fault of my python wrapper's custom playback_callback slowing everything down while it tries to process the stream of non-note events sent by `fluid_track_send_events` to get the song state correct at the seek point. Poor old Python can't keep up, but the C version seeks perfectly with no audio artifacts that I can hear - I included some code and sample recordings below. I used FluidR3_GM.sf2 for the soundfont, but omitted it from the zip for space.

[fluid_player_seek_test.zip](https://github.com/FluidSynth/fluidsynth/files/7190516/fluid_player_seek_test.zip)

If I use jack audio with the python wrapper, the xruns produce a buzz - if seeking to earlier in the song it sounded like a click. Using alsa for audio just produces a gap in the audio. I think I can fix my python code by unsetting the player's playback_callback while seeking, but it's not an issue with Fluidsynth.

**Reply by derselbst** [reply]
Ok, good to know. Still, thanks for the example tunes!

Your changes in #980 look pretty good to me already. IMO, we don't need the `player.global-mute` setting - I think we can safely replace the old global `fluid_synth_all_sounds_off()` with the new CC `all_sounds_off`.

**Reply by albedozero** [reply]
Great, the latest commit does that. I'm sending `ALL_SOUND_OFF` when both seeking and stopping, but the current code calls `fluid_synth_all_notes_off()` when stopping, which can cause some hanging notes if sustain/sostenuto is active. Is there a reason to send an `ALL_NOTES_OFF` CC when stopping instead?

**Reply by derselbst** [reply]
> Is there a reason to send an ALL_NOTES_OFF CC when stopping instead?

I think it's because the `fluid_player_stop` is actually a misnamed "pause" function. And when pressing the pause button it seemed more natural for a MIDI player to send out `ALL_NOTES_OFF`. I have no strong preference though. We could probably live with `ALL_SOUND_OFF` as well. @mawe42 ?

**Reply by mawe42** [reply]
If `fluid_player_stop` is really just a pause, and if the recommended way to implement a proper stop is to `fluid_player_stop(); fluid_player_seek(0)`, then I would vote to leave stop as `ALL_NOTES_OFF` and make seek use `ALL_SOUNDS_OFF` **and** `ALL_CTRL_OFF`.

**Reply by albedozero** [reply]
> If fluid_player_stop is really just a pause, and if the recommended way to implement a proper stop is to fluid_player_stop(); fluid_player_seek(0), then I would vote to leave stop as ALL_NOTES_OFF

Even if `fluid_player_stop` is a pause, I feel like most users [expect it should halt sound from the player](https://github.com/FluidSynth/fluidsynth/discussions/958#discussioncomment-1358654), even if the sustain pedal is active at that moment, without having to additionally call `fluid_player_seek(0)`. This setup also makes it cumbersome if a user really does just want to pause playback and recommence from the same spot - they have to call `fluid_player_stop`, record the current tick position, call `fluid_player_seek(0)` to silence any sustained notes, then to recommence call `fluid_player_seek(tick)` to get the player back where it was and finally `fluid_player_play()`.

> and make seek use ALL_SOUNDS_OFF and ALL_CTRL_OFF.

@mawe42 I think you suggested `ALL_CTRL_OFF` above as a way to solve hanging notes in case c) where a player did a seek from a point where the sustain pedal is active to a point earlier in the song where it has never been pressed, and `fluid_track_send_events` thus has no CC64=0 event to process. But, the code I was testing was sending notes_off when seeking - if we're sending `ALL_SOUND_OFF` then controls_off shouldn't be needed to solve that problem. In fact, it might reset controllers that the user has set for specific reasons - I'm thinking of `MIDI_VOLUME` in particular, but there could be others.

So, I feel like sending `ALL_SOUND_OFF` when stopping produces the more expected behavior, and refraining from sending `ALL_CTRL_OFF` when seeking gives the user more flexibility.

**Reply by albedozero** [reply]
Another problem I realized with `ALL_CTRL_OFF` is that when seeking forward this would change the current state of the player, so `fluid_track_send_events()` would have to be rewritten to read from the beginning of the track whether seeking forward or backward.

Let me know what you think either way - I won't argue any more 🙂

**Reply by derselbst** [reply]
> So, I feel like sending ALL_SOUND_OFF when stopping produces the more expected behavior, and refraining from sending ALL_CTRL_OFF when seeking gives the user more flexibility.

Agreed. Let's do it that way: `ALL_SOUND_OFF` for both: stop and seek. And no `ALL_CTRL_OFF` for now. That should be good enough. I'll give your PR a closer review this weekend.
